crates/red_knot_python_semantic/resources/mdtest/descriptor_protocol.md

```diff
@@ -186,13 +252,165 @@ class Descriptor:
             return "called on class object"
 
 class C:
-    d = Descriptor()
+    d: Descriptor = Descriptor()
 
 # TODO: should be `Literal["called on class object"]
-reveal_type(C.d)  # revealed: Unknown | Descriptor
+reveal_type(C.d)  # revealed: LiteralString
 
 # TODO: should be `Literal["called on instance"]
-reveal_type(C().d)  # revealed: Unknown | Descriptor
+reveal_type(C().d)  # revealed: LiteralString
+```
+
+## Undeclared descriptor arguments
+
+If a descriptor attribute is not declared, we union with `Unknown`, just like for regular
+attributes, since that attribute could be overwritten externally. Even data-descriptors with a
+`__set__` method can be overwritten when accessed through a class object.
+
+```py
+class Descriptor:
+    def __get__(self, instance: object, owner: type | None = None) -> int:
+        return 1
```

Comment:
I know this specific example is showing what happens if a descriptor is _set_ on the class object. But the `__get__` signature here implies that if the attribute is _read_ (rather than _set_) on the class, we should emit an error, since `None` would be passed as the first argument to `__get__` in that case. Is that correct?

Reply:
> But the `__get__` signature here implies that if the attribute is _read_ (rather than _set_) on the class, we should emit an error, since `None` would be passed as the first argument to `__get__` in that case. Is that correct?

I you access `C.descriptor` on the class itself, `None` would be passed as the `instance` argument, yes. But that should not produce an error. `None` is compatible with `object`. This `__get__` method here was not supposed to distinguish between class-based or instance-based access.

Reply:
> `None` is compatible with `object`

duh, of course. Sorry, I guess I forgot this was the case :-)